File: fl_mlkit_scanning/android/src/main/kotlin/fl/mlkit/scanning/FlMlKitScanningPlugin.kt
package fl.mlkit.scanning

import android.annotation.SuppressLint
import android.graphics.BitmapFactory
import android.graphics.Point
import android.graphics.Rect
import androidx.camera.core.ImageProxy
import com.google.mlkit.vision.barcode.BarcodeScanner
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.common.InputImage
import fl.camera.FlCamera
import fl.channel.FlDataStreamHandlerCancel
import fl.channel.FlEvent
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class FlMlKitScanningPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {

    private lateinit var channel: MethodChannel
    private var options = BarcodeScannerOptions.Builder().setBarcodeFormats(Barcode.FORMAT_ALL_FORMATS).build()
    private var scanner: BarcodeScanner? = null
    private var lastCurrentTime = 0L

    private var imageProxyHandler: FlDataStreamHandlerCancel? = null

    override fun onAttachedToEngine(pluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(pluginBinding.binaryMessenger, "fl.mlkit.scanning")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    @SuppressLint("UnsafeOptInUsageError")
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "setParams" -> {
                val frequency = call.argument<Int>("frequency")!!.toLong()
                val canScanning = call.argument<Boolean>("canScanning")!!
                imageProxyHandler?.invoke()
                imageProxyHandler = null
                imageProxyHandler = FlCamera.flDataStream.listen { imageProxy ->
                    val mediaImage = imageProxy.image
                    val currentTime = System.currentTimeMillis()
                    if (currentTime - lastCurrentTime >= frequency && mediaImage != null && canScanning) {
                        val inputImage = InputImage.fromMediaImage(mediaImage, imageProxy.imageInfo.rotationDegrees)
                        analysis(inputImage, null, imageProxy)
                        lastCurrentTime = currentTime
                    } else {
                        imageProxy.close()
                    }
                }
                result.success(true)
            }

            "setBarcodeFormat" -> {
                setBarcodeFormat(call)
                scanner?.close()
                scanner = null
                result.success(true)
            }

            "scanningImageByte" -> scanningImageByte(call, result)
            "dispose" -> {
                imageProxyHandler?.invoke()
                imageProxyHandler = null
                scanner?.close()
                scanner = null
                result.success(true)
            }

            else -> result.notImplemented()
        }
    }

    private fun scanningImageByte(call: MethodCall, result: MethodChannel.Result) {
        val useEvent = call.argument<Boolean>("useEvent")!!
        val byteArray = call.argument<ByteArray>("byte")!!
        var rotationDegrees = call.argument<Int>("rotationDegrees")
        val bitmap = BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size)
        if (bitmap == null) {
            result.success(null)
            return
        }
        if (rotationDegrees == null) rotationDegrees = 0
        val inputImage = InputImage.fromBitmap(bitmap, rotationDegrees)
        analysis(inputImage, if (useEvent) null else result, null)
    }

    private fun setBarcodeFormat(call: MethodCall) {
        val barcodeFormats = call.arguments as List<*>
        val builder = BarcodeScannerOptions.Builder()
        if (barcodeFormats.isNotEmpty()) {
            val formats = barcodeFormats.map { type -> getBarcodeFormat(type as String) }
            builder.setBarcodeFormats(formats.first(), *formats.toIntArray())
        } else {
            builder.setBarcodeFormats(Barcode.FORMAT_ALL_FORMATS, Barcode.FORMAT_CODE_39)
        }
        options = builder.build()
    }

    private fun getBarcodeFormat(type: String): Int {
        when (type) {
            "unknown" -> return Barcode.FORMAT_UNKNOWN
            "all" -> return Barcode.FORMAT_ALL_FORMATS
            "code128" -> return Barcode.FORMAT_CODE_128
            "code39" -> return Barcode.FORMAT_CODE_39
            "code93" -> return Barcode.FORMAT_CODE_93
            "codaBar" -> return Barcode.FORMAT_CODABAR
            "dataMatrix" -> return Barcode.FORMAT_DATA_MATRIX
            "ean13" -> return Barcode.FORMAT_EAN_13
            "ean8" -> return Barcode.FORMAT_EAN_8
            "itf" -> return Barcode.FORMAT_ITF
            "qrCode" -> return Barcode.FORMAT_QR_CODE
            "upcA" -> return Barcode.FORMAT_UPC_A
            "upcE" -> return Barcode.FORMAT_UPC_E
            "pdf417" -> return Barcode.FORMAT_PDF417
            "aztec" -> return Barcode.FORMAT_AZTEC
        }
        return Barcode.FORMAT_UNKNOWN
    }


    private fun analysis(inputImage: InputImage, result: MethodChannel.Result?, imageProxy: ImageProxy?) {
        val list: ArrayList<Map<String, Any?>> = ArrayList()
        getBarcodeScanner().process(inputImage).addOnSuccessListener { barcodes ->
            for (barcode in barcodes) {
                list.add(barcode.data)
            }
            var width = inputImage.width
            var height = inputImage.height
            if (width > height) {
                width -= height
                height += width
                width = height - width
            }

            val map = mapOf("height" to height.toDouble(), "width" to width.toDouble(), "barcodes" to list)
            if (result == null) {
                if (list.isNotEmpty()) {
                    FlEvent.send(map)
                }
            } else {
                result.success(map)
            }
        }.addOnFailureListener { result?.success(null) }.addOnCompleteListener { imageProxy?.close() }
    }

    private fun getBarcodeScanner(): BarcodeScanner {
        if (scanner == null) {
            scanner = BarcodeScanning.getClient(options)
        }
        return scanner!!
    }


    private val Barcode.data: Map<String, Any?>
        get() = mapOf(
                "value" to rawValue,
                "type" to valueType,
                "corners" to cornerPoints?.map { corner -> corner.data },
                "boundingBox" to boundingBox?.data,
                "displayValue" to displayValue,
                "format" to format,
                "calendarEvent" to calendarEvent?.data,
                "contactInfo" to contactInfo?.data,
                "driverLicense" to driverLicense?.data,
                "email" to email?.data,
                "geoPoint" to geoPoint?.data,
                "phone" to phone?.data,
                "sms" to sms?.data,
                "url" to url?.data,
                "wifi" to wifi?.data,
                "bytes" to rawBytes,
        )
    private val Rect.data: Map<String, Int>
        get() = mapOf("top" to top, "bottom" to bottom, "left" to left, "right" to right, "width" to width(), "height" to height())

    private val Point.data: Map<String, Double>
        get() = mapOf("x" to x.toDouble(), "y" to y.toDouble())

    private val Barcode.CalendarEvent.data: Map<String, Any?>
        get() = mapOf("description" to description, "end" to end?.rawValue, "location" to location, "organizer" to organizer, "start" to start?.rawValue, "status" to status, "summary" to summary)

    private val Barcode.ContactInfo.data: Map<String, Any?>
        get() = mapOf("addresses" to addresses.map { address -> address.data }, "emails" to emails.map { email -> email.data }, "name" to name?.data, "organization" to organization, "phones" to phones.map { phone -> phone.data }, "title" to title, "urls" to urls)

    private val Barcode.Address.data: Map<String, Any?>
        get() = mapOf("addressLines" to addressLines, "type" to type)

    private val Barcode.PersonName.data: Map<String, Any?>
        get() = mapOf("first" to first, "formattedName" to formattedName, "last" to last, "middle" to middle, "prefix" to prefix, "pronunciation" to pronunciation, "suffix" to suffix)

    private val Barcode.DriverLicense.data: Map<String, Any?>
        get() = mapOf("addressCity" to addressCity, "addressState" to addressState, "addressStreet" to addressStreet, "addressZip" to addressZip, "birthDate" to birthDate, "documentType" to documentType, "expiryDate" to expiryDate, "firstName" to firstName, "gender" to gender, "issueDate" to issueDate, "issuingCountry" to issuingCountry, "lastName" to lastName, "licenseNumber" to licenseNumber, "middleName" to middleName)

    private val Barcode.Email.data: Map<String, Any?>
        get() = mapOf("address" to address, "body" to body, "subject" to subject, "type" to type)

    private val Barcode.GeoPoint.data: Map<String, Any?>
        get() = mapOf("latitude" to lat, "longitude" to lng)

    private val Barcode.Phone.data: Map<String, Any?>
        get() = mapOf("number" to number, "type" to type)

    private val Barcode.Sms.data: Map<String, Any?>
        get() = mapOf("message" to message, "phoneNumber" to phoneNumber)

    private val Barcode.UrlBookmark.data: Map<String, Any?>
        get() = mapOf("title" to title, "url" to url)

    private val Barcode.WiFi.data: Map<String, Any?>
        get() = mapOf("encryptionType" to encryptionType, "password" to password, "ssid" to ssid)
}

File: fl_mlkit_scanning/ios/Classes/FlMlKitScanningPlugin.swift
import AVFoundation
import fl_camera
import fl_channel
import Flutter
import Foundation
import MLKitBarcodeScanning
import MLKitVision
import UIKit

public class FlMlKitScanningPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?

    private var options = BarcodeScannerOptions(formats: .all)
    private var lastCurrentTime: TimeInterval = 0
    private var scanner: BarcodeScanner?
    private var bufferHandler: FlDataStreamHandlerCancel?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "fl.mlkit.scanning",
                                           binaryMessenger: registrar.messenger())
        let instance = FlMlKitScanningPlugin(channel)
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    init(_ channel: FlutterMethodChannel) {
        self.channel = channel
        super.init()
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "setParams":
            let arguments = call.arguments as! [AnyHashable: Any?]
            let frequency = arguments["frequency"] as! Double
            let canScanning = arguments["canScanning"] as! Bool
            bufferHandler?()
            bufferHandler = nil
            bufferHandler = FlCamera.shared.flDataStream.listen { [self] buffer in
                let currentTime = Date().timeIntervalSince1970 * 1000
                if currentTime - lastCurrentTime >= frequency, canScanning {
                    let buffer = CMSampleBufferGetImageBuffer(buffer)
                    analysis(buffer!.image, nil)
                    lastCurrentTime = currentTime
                }
            }
            result(true)
        case "setBarcodeFormat":
            setBarcodeFormat(call)
            scanner = nil
            result(true)
        case "scanningImageByte":
            let arguments = call.arguments as! [AnyHashable: Any?]
            let useEvent = arguments["useEvent"] as! Bool
            let uint8list = arguments["byte"] as! FlutterStandardTypedData?
            if uint8list != nil {
                let image = UIImage(data: uint8list!.data)
                if image != nil {
                    analysis(image!, useEvent ? nil : result)
                    return
                }
            }
            result(nil)
        case "dispose":
            scanner = nil
            bufferHandler?()
            bufferHandler = nil
            result(true)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        channel?.setMethodCallHandler(nil)
        channel = nil
    }

    func setBarcodeFormat(_ call: FlutterMethodCall) {
        let barcodeFormats = call.arguments as! [String]
        if !barcodeFormats.isEmpty {
            var formats = BarcodeFormat()
            for barcodeFormat in barcodeFormats {
                switch barcodeFormat {
                case "unknown":
                    formats.insert(.all)
                case "all":
                    formats.insert(.all)
                case "code128":
                    formats.insert(.code128)
                case "code39":
                    formats.insert(.code39)
                case "code93":
                    formats.insert(.code93)
                case "codaBar":
                    formats.insert(.codaBar)
                case "dataMatrix":
                    formats.insert(.dataMatrix)
                case "ean13":
                    formats.insert(.EAN13)
                case "ean8":
                    formats.insert(.EAN8)
                case "itf":
                    formats.insert(.ITF)
                case "qrCode":
                    formats.insert(.qrCode)
                case "upcA":
                    formats.insert(.UPCA)
                case "upcE":
                    formats.insert(.UPCE)
                case "pdf417":
                    formats.insert(.PDF417)
                case "aztec":
                    formats.insert(.aztec)
                default:
                    break
                }
            }
            options = BarcodeScannerOptions(formats: formats)
        }
    }

    private func getBarcodeScanner() -> BarcodeScanner {
        if scanner == nil {
            scanner = BarcodeScanner.barcodeScanner(options: options)
        }
        return scanner!
    }

    func analysis(_ image: UIImage, _ result: FlutterResult?) {
        let visionImage = VisionImage(image: image)
        if FlCamera.shared.cameraTexture == nil {
            visionImage.orientation = .up
        } else {
            visionImage.orientation = FlCamera.shared.cameraTexture!.imageOrientation()
        }
        getBarcodeScanner().process(visionImage) { barcodes, error in
            if error == nil, barcodes != nil {
                var list = [[String: Any?]]()
                for barcode in barcodes! {
                    list.append(barcode.data)
                }
                let map = [
                    "height": image.size.height,
                    "width": image.size.width,
                    "barcodes": list
                ] as [String: Any?]
                if result == nil {
                    if !list.isEmpty {
                        FlEvent.shared.send(map)
                    }
                } else {
                    result!(map)
                }
            }
        }
    }
}

extension CVBuffer {
    var image: UIImage {
        let ciImage = CIImage(cvPixelBuffer: self)
        let cgImage = CIContext().createCGImage(ciImage, from: ciImage.extent)
        return UIImage(cgImage: cgImage!)
    }

    var image1: UIImage {
        // Lock the base address of the pixel buffer
        CVPixelBufferLockBaseAddress(self, CVPixelBufferLockFlags.readOnly)
        // Get the number of bytes per row for the pixel buffer
        let baseAddress = CVPixelBufferGetBaseAddress(self)
        // Get the number of bytes per row for the pixel buffer
        let bytesPerRow = CVPixelBufferGetBytesPerRow(self)
        // Get the pixel buffer width and height
        let width = CVPixelBufferGetWidth(self)
        let height = CVPixelBufferGetHeight(self)
        // Create a device-dependent RGB color space
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        // Create a bitmap graphics context with the sample buffer data
        var bitmapInfo = CGBitmapInfo.byteOrder32Little.rawValue
        bitmapInfo |= CGImageAlphaInfo.premultipliedFirst.rawValue & CGBitmapInfo.alphaInfoMask.rawValue
        // let bitmapInfo: UInt32 = CGBitmapInfo.alphaInfoMask.rawValue
        let context = CGContext(data: baseAddress, width: width, height: height, bitsPerComponent: 8, bytesPerRow: bytesPerRow, space: colorSpace, bitmapInfo: bitmapInfo)
        // Create a Quartz image from the pixel data in the bitmap graphics context
        let quartzImage = context?.makeImage()
        // Unlock the pixel buffer
        CVPixelBufferUnlockBaseAddress(self, CVPixelBufferLockFlags.readOnly)
        // Create an image object from the Quartz image
        return UIImage(cgImage: quartzImage!)
    }
}

extension CGRect {
    var data: [String: Any?] {
        [
            "x": origin.x,
            "y": origin.y,
            "width": width,
            "height": height
        ]
    }
}

extension Barcode {
    var data: [String: Any?] {
        ["corners": cornerPoints?.map {
            $0.cgPointValue.data
        },
        "format": format.rawValue,
        "bytes": rawData,
        "value": rawValue,
        "type": valueType.rawValue,
        "calendarEvent": calendarEvent?.data,
        "contactInfo": contactInfo?.data,
        "driverLicense": driverLicense?.data,
        "email": email?.data,
        "geoPoint": geoPoint?.data,
        "phone": phone?.data,
        "sms": sms?.data,
        "url": url?.data,
        "wifi": wifi?.data,
        "displayValue": displayValue,
        "boundingBox": frame.data]
    }
}

extension CGPoint {
    var data: [String: Any?] {
        ["x": NSNumber(value: x.native), "y": NSNumber(value: y.native)]
    }
}

extension BarcodeCalendarEvent {
    var data: [String: Any?] {
        ["description": eventDescription, "end": end?.rawValue, "location": location, "organizer": organizer, "start": start?.rawValue, "status": status, "summary": summary]
    }
}

extension Date {
    var rawValue: String {
        ISO8601DateFormatter().string(from: self)
    }
}

extension BarcodeContactInfo {
    var data: [String: Any?] {
        ["addresses": addresses?.map {
            $0.data
        }, "emails": emails?.map {
            $0.data
        }, "name": name?.data, "organization": organization, "phones": phones?.map {
            $0.data
        }, "title": jobTitle, "urls": urls]
    }
}

extension BarcodeAddress {
    var data: [String: Any?] {
        ["addressLines": addressLines, "type": type.rawValue]
    }
}

extension BarcodePersonName {
    var data: [String: Any?] {
        ["first": first, "formattedName": formattedName, "last": last, "middle": middle, "prefix": prefix, "pronunciation": pronunciation, "suffix": suffix]
    }
}

extension BarcodeDriverLicense {
    var data: [String: Any?] {
        ["addressCity": addressCity, "addressState": addressState, "addressStreet": addressStreet, "addressZip": addressZip, "birthDate": birthDate, "documentType": documentType, "expiryDate": expiryDate, "firstName": firstName, "gender": gender, "issueDate": issuingDate, "issuingCountry": issuingCountry, "lastName": lastName, "licenseNumber": licenseNumber, "middleName": middleName]
    }
}

extension BarcodeEmail {
    var data: [String: Any?] {
        ["address": address, "body": body, "subject": subject, "type": type.rawValue]
    }
}

extension BarcodeGeoPoint {
    var data: [String: Any?] {
        ["latitude": latitude, "longitude": longitude]
    }
}

extension BarcodePhone {
    var data: [String: Any?] {
        ["number": number, "type": type.rawValue]
    }
}

extension BarcodeSMS {
    var data: [String: Any?] {
        ["message": message, "phoneNumber": phoneNumber]
    }
}

extension BarcodeURLBookmark {
    var data: [String: Any?] {
        ["title": title, "url": url]
    }
}

extension BarcodeWifi {
    var data: [String: Any?] {
        ["encryptionType": type.rawValue, "password": password, "ssid": ssid]
    }
}
